provider/docker/docker.go

```diff
@@ -259,44 +259,80 @@ func (p *Provider) Provide(configurationChan chan<- types.ConfigMessage, pool *s
-		"getServiceEntryPoints":       p.getServiceEntryPoints,
-		"getServiceBasicAuth":         p.getServiceBasicAuth,
-		"getServiceFrontendRule":      p.getServiceFrontendRule,
-		"getServicePassHostHeader":    p.getServicePassHostHeader,
-		"getServicePriority":          p.getServicePriority,
-		"getServiceBackend":           p.getServiceBackend,
-		"getWhitelistSourceRange":     p.getWhitelistSourceRange,
-		"getRequestHeaders":           p.getRequestHeaders,
-		"getResponseHeaders":          p.getResponseHeaders,
-		"hasRequestHeaders":           p.hasRequestHeaders,
-		"hasResponseHeaders":          p.hasResponseHeaders,
+		"getBackend":                        p.getBackend,
+		"getIPAddress":                      p.getIPAddress,
+		"getPort":                           p.getPort,
+		"getWeight":                         p.getWeight,
+		"getDomain":                         p.getDomain,
+		"getProtocol":                       p.getProtocol,
+		"getPassHostHeader":                 p.getPassHostHeader,
+		"getPriority":                       p.getPriority,
+		"getEntryPoints":                    p.getEntryPoints,
+		"getBasicAuth":                      p.getBasicAuth,
+		"getFrontendRule":                   p.getFrontendRule,
+		"hasCircuitBreakerLabel":            p.hasCircuitBreakerLabel,
+		"getCircuitBreakerExpression":       p.getCircuitBreakerExpression,
+		"hasLoadBalancerLabel":              p.hasLoadBalancerLabel,
+		"getLoadBalancerMethod":             p.getLoadBalancerMethod,
+		"hasMaxConnLabels":                  p.hasMaxConnLabels,
+		"getMaxConnAmount":                  p.getMaxConnAmount,
+		"getMaxConnExtractorFunc":           p.getMaxConnExtractorFunc,
+		"getSticky":                         p.getSticky,
+		"getStickinessCookieName":           p.getStickinessCookieName,
+		"hasStickinessLabel":                p.hasStickinessLabel,
+		"getIsBackendLBSwarm":               p.getIsBackendLBSwarm,
+		"hasServices":                       p.hasServices,
+		"getServiceNames":                   p.getServiceNames,
+		"getServicePort":                    p.getServicePort,
+		"getServiceWeight":                  p.getServiceWeight,
+		"getServiceProtocol":                p.getServiceProtocol,
+		"getServiceEntryPoints":             p.getServiceEntryPoints,
+		"getServiceBasicAuth":               p.getServiceBasicAuth,
+		"getServiceFrontendRule":            p.getServiceFrontendRule,
+		"getServicePassHostHeader":          p.getServicePassHostHeader,
+		"getServicePriority":                p.getServicePriority,
+		"getServiceBackend":                 p.getServiceBackend,
+		"getWhitelistSourceRange":           p.getWhitelistSourceRange,
+		"hasRequestHeaders":                 p.hasRequestHeaders,
+		"getRequestHeaders":                 p.getRequestHeaders,
+		"hasResponseHeaders":                p.hasResponseHeaders,
+		"getResponseHeaders":                p.getResponseHeaders,
+		"hasAllowedHostsHeaders":            p.hasAllowedHostsHeaders,
```

Comment:
You can simplify your code like that:
```
"hasAllowedHostsHeaders":            p.hasLabel(types.LabelFrontendAllowedHosts),
```

```go
func (p Provider) hasLabel(label string) func(container dockerData) bool {
	return func(container dockerData) bool {
		label, err := getLabel(container, label)
		return err == nil && len(label) > 0
	}
}
```

Reply:
Good Call!

Was just getting code roughed in. I knew that there is a lot of duplication, but I wanted to see it as a whole first, before condensing :)